gio/src/socket.rs

```diff
@@ -61,6 +69,185 @@ impl AsRawSocket for Socket {
+        control_messages: Option<&'m mut SocketControlMessages>,
+    ) -> Self {
+        let address = address
+            .as_mut()
+            .map(|a| {
+                assert!(a.is_none());
+                *a as *mut _ as *mut _
+            })
+            .unwrap_or_else(ptr::null_mut);
+        let (control_messages, num_control_messages) = control_messages
+            .map(|c| (&mut c.ptr as *mut _, &mut c.len as *mut _))
+            .unwrap_or_else(|| (ptr::null_mut(), ptr::null_mut()));
+        Self {
+            message: ffi::GInputMessage {
+                address,
+                vectors: vectors.as_mut_ptr() as *mut ffi::GInputVector,
+                num_vectors: vectors.len().try_into().unwrap(),
+                bytes_received: 0,
+                flags: 0,
+                control_messages,
+                num_control_messages,
+            },
+            address: PhantomData,
+            vectors: PhantomData,
+            control_messages: PhantomData,
+        }
+    }
+    #[inline]
+    pub const fn flags(&self) -> i32 {
+        self.message.flags
+    }
+    #[inline]
+    pub const fn bytes_received(&self) -> usize {
+        self.message.bytes_received
+    }
+}
+
+#[doc(alias = "GOutputVector")]
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct OutputVector<'v> {
+    _vector: ffi::GOutputVector,
+    buffer: PhantomData<&'v [u8]>,
+}
+
+impl<'v> OutputVector<'v> {
+    #[inline]
+    pub const fn new(buffer: &'v [u8]) -> Self {
+        Self {
+            _vector: ffi::GOutputVector {
+                buffer: buffer.as_ptr() as *const _,
+                size: buffer.len(),
+            },
+            buffer: PhantomData,
+        }
+    }
+}
+
+#[doc(alias = "GOutputMessage")]
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct OutputMessage<'m> {
+    message: ffi::GOutputMessage,
+    address: PhantomData<Option<&'m SocketAddress>>,
+    vectors: PhantomData<&'m [OutputVector<'m>]>,
+    control_messages: PhantomData<&'m [SocketControlMessage]>,
+}
+
+impl<'m> OutputMessage<'m> {
+    pub fn new<A: IsA<SocketAddress>>(
+        address: Option<&'m A>,
+        vectors: &'m [OutputVector<'m>],
+        control_messages: &'m [SocketControlMessage],
+    ) -> Self {
+        Self {
+            message: ffi::GOutputMessage {
+                address: address
+                    .map(|a| a.upcast_ref::<SocketAddress>().as_ptr())
+                    .unwrap_or_else(ptr::null_mut),
+                vectors: mut_override(vectors.as_ptr() as *const ffi::GOutputVector),
+                num_vectors: vectors.len().try_into().unwrap(),
+                bytes_sent: 0,
+                control_messages: control_messages.as_ptr() as *mut _,
+                num_control_messages: control_messages.len().try_into().unwrap(),
+            },
+            address: PhantomData,
+            vectors: PhantomData,
+            control_messages: PhantomData,
+        }
+    }
+    #[inline]
```

Comment:
Why do you put `#[inline]` on everything everywhere btw? :) Does it actually make any difference here?

Generally the LLVM heuristic for inlining is "yes", so in the majority of cases this has no effect at all.

Reply:
No particular reason, I noticed some places we have it and some places we don't